cc/algorithms/partition-selection.h

```diff
@@ -0,0 +1,327 @@
+#include "base/status_macros.h"
+#include "base/canonical_errors.h"
+#include "algorithms/numerical-mechanisms.h"
+#include "algorithms/rand.h"
+#include <iostream>
+
+namespace differential_privacy{
+
+// Provides a common abstraction for PartitionSelectionStrategy. Each partition
+// selection strategy class has a builder with which it can be instantiated, and
+// calling ShouldKeep will return true if a partition with the given number of
+// users should be kept based on the values the partition selection strategy
+// was instantiated with (while ShouldKeep will return false if the partition
+// should have been dropped).
+class PartitionSelectionStrategy {
+ public:
+  //Builder base class
+  class Builder {
+   public:
+     virtual ~Builder() = default;
+
+     Builder& SetEpsilon(double epsilon) {
+       epsilon_ = epsilon;
+       return *this;
+     }
+
+     Builder& SetDelta(double delta) {
+       delta_ = delta;
+       return *this;
+     }
+
+     Builder& SetMaxPartitionsContributed(int max_partitions_contributed) {
+       max_partitions_contributed_ = max_partitions_contributed;
+       return *this;
+     }
+
+     virtual base::StatusOr<std::unique_ptr<PartitionSelectionStrategy>>
+     Build() = 0;
+
+   protected:
+     // Checks if the max number of partitions contributed  to is set and valid.
+     base::Status MaxPartitionsContributedIsSetAndValid() {
+       if (!max_partitions_contributed_.has_value()) {
+         return base::InvalidArgumentError(
+          "Max number of partitions a user can contribute to has to be set.");
+       }
+       if (max_partitions_contributed_ <= 0) {
+         return base::InvalidArgumentError(absl::StrCat(
+           "Max number of partitions a user can contribute to has to be"
+           " positive but is ",
+           delta_.value()));
+       }
+       return base::OkStatus();
+     }
+
+     // Checks if delta is set and valid.
+     base::Status DeltaIsSetAndValid() {
+       if (!delta_.has_value()) {
+         return base::InvalidArgumentError("Delta has to be set.");
+       }
+       if (!std::isfinite(delta_.value())) {
+         return base::InvalidArgumentError(
+           absl::StrCat("Delta has to be finite but is ", delta_.value()));
+       }
+       if (delta_.value() <= 0 || delta_.value() >= 1) {
+         return base::InvalidArgumentError(absl::StrCat(
+           "Delta has to be in the interval (0,1) but is ",
+           delta_.value()));
+        }
+        return base::OkStatus();
+      }
+
+     absl::optional<double> GetEpsilon() { return epsilon_; }
+
+     absl::optional<double> GetDelta() { return delta_; }
+
+     absl::optional<int> GetMaxPartitionsContributed() {
+       return max_partitions_contributed_;
+     }    
+
+    private:
+     absl::optional<double> epsilon_;
+     absl::optional<double> delta_;
+     absl::optional<int> max_partitions_contributed_;
+  };
+
+  PartitionSelectionStrategy(double epsilon, double delta,
```

Comment:
Make the constructor protected to force folks to construct it through the builder.